File: internal/compute/store.go
package compute

import (
	"container/list"
	"context"
	"encoding/json"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"

	"pii/internal/models"
)

// Store is the correspondence store with in-memory LRU+TTL cache and Redis write-through.
type Store struct {
	rdb    *redis.Client
	ns     string
	ttl    time.Duration
	cache  *lruCache
	encKey []byte
	mu     sync.Mutex
	hits   uint64
	misses uint64
}

// NewStore creates a correspondence store.
func NewStore(rdb *redis.Client, ns string, ttl time.Duration, cacheMax int, encKey []byte) *Store {
	return &Store{
		rdb:    rdb,
		ns:     ns,
		ttl:    ttl,
		cache:  newLRUCache(cacheMax),
		encKey: encKey,
	}
}

// Get returns a correspondence record by payload id.
func (s *Store) Get(ctx context.Context, payloadID string) (*models.CorrRecord, error) {
	if rec, ok := s.cache.get(payloadID); ok {
		s.recordHit()
		return rec, nil
	}
	s.recordMiss()
	if s.rdb == nil {
		return nil, nil
	}
	key := corrKey(s.ns, payloadID)
	raw, err := s.rdb.Get(ctx, key).Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	plain, err := decrypt([]byte(raw), s.encKey)
	if err != nil {
		return nil, err
	}
	var rec models.CorrRecord
	if err := json.Unmarshal(plain, &rec); err != nil {
		return nil, err
	}
	s.cache.put(payloadID, &rec)
	return &rec, nil
}

// Put stores a correspondence record (write-through: cache + Redis with TTL).
func (s *Store) Put(ctx context.Context, payloadID, original, mask string, types []string) error {
	rec := &models.CorrRecord{
		Original:  original,
		Mask:      mask,
		Types:     types,
		CreatedTS: float64(time.Now().Unix()),
	}
	plain, err := json.Marshal(rec)
	if err != nil {
		return err
	}
	enc, err := encrypt(plain, s.encKey)
	if err != nil {
		return err
	}
	key := corrKey(s.ns, payloadID)
	if s.rdb != nil {
		if err := s.rdb.Set(ctx, key, enc, s.ttl).Err(); err != nil {
			return err
		}
	}
	s.cache.put(payloadID, rec)
	return nil
}

// Lookup resolves the direction: original→mask or mask→original.
func (s *Store) Lookup(ctx context.Context, payloadID, payload string) (string, Direction, bool, error) {
	rec, err := s.Get(ctx, payloadID)
	if err != nil || rec == nil {
		return "", DirectionMask, false, err
	}
	if payload == rec.Original {
		return rec.Mask, DirectionMask, true, nil
	}
	if payload == rec.Mask {
		return rec.Original, DirectionDemask, true, nil
	}
	return "", DirectionMask, false, nil
}

// Size returns the number of cached records.
func (s *Store) Size() int {
	return s.cache.len()
}

func (s *Store) recordHit() {
	s.mu.Lock()
	s.hits++
	s.mu.Unlock()
}

func (s *Store) recordMiss() {
	s.mu.Lock()
	s.misses++
	s.mu.Unlock()
}

// CacheHitRate returns the in-memory cache hit ratio.
func (s *Store) CacheHitRate() float64 {
	s.mu.Lock()
	defer s.mu.Unlock()
	total := s.hits + s.misses
	if total == 0 {
		return 0
	}
	return float64(s.hits) / float64(total)
}

type lruEntry struct {
	key   string
	value *models.CorrRecord
	exp   time.Time
}

type lruCache struct {
	mu    sync.Mutex
	max   int
	ll    *list.List
	items map[string]*list.Element
}

func newLRUCache(max int) *lruCache {
	if max <= 0 {
		max = 100000
	}
	return &lruCache{
		max:   max,
		ll:    list.New(),
		items: map[string]*list.Element{},
	}
}

func (c *lruCache) get(key string) (*models.CorrRecord, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	el, ok := c.items[key]
	if !ok {
		return nil, false
	}
	entry := el.Value.(*lruEntry)
	if time.Now().After(entry.exp) {
		c.remove(el)
		return nil, false
	}
	c.ll.MoveToFront(el)
	return entry.value, true
}

func (c *lruCache) put(key string, rec *models.CorrRecord) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if el, ok := c.items[key]; ok {
		el.Value.(*lruEntry).value = rec
		el.Value.(*lruEntry).exp = time.Now().Add(24 * time.Hour)
		c.ll.MoveToFront(el)
		return
	}
	entry := &lruEntry{key: key, value: rec, exp: time.Now().Add(24 * time.Hour)}
	el := c.ll.PushFront(entry)
	c.items[key] = el
	if c.ll.Len() > c.max {
		c.remove(c.ll.Back())
	}
}

func (c *lruCache) remove(el *list.Element) {
	entry := el.Value.(*lruEntry)
	delete(c.items, entry.key)
	c.ll.Remove(el)
}

func (c *lruCache) len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.ll.Len()
}

File: compute/src/store.rs
//! Хранилище соответствий «оригинал ↔ маска».
//!
//! Обеспечивает идемпотентное демаскирование: по идентификатору полезной
//! нагрузки хранится запись соответствия, позволяющая восстановить
//! оригинал из маски и наоборот. Используется LRU-кэш с TTL поверх Redis.

use std::num::NonZeroUsize;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use lru::LruCache;
use redis::aio::ConnectionManager;
use redis::AsyncCommands;

use crate::keys::corr_key;
use crate::models::CorrRecord;

/// Тип результата операций хранилища.
pub type Result<T> = std::result::Result<T, redis::RedisError>;

/// Внутреннее состояние LRU-кэша записей соответствия.
struct CacheInner {
    /// LRU-кэш «payload_id → (запись, время записи)».
    entries: LruCache<String, (CorrRecord, Instant)>,
}

/// Хранилище соответствий с in-memory LRU+TTL кэшем.
#[derive(Clone)]
pub struct CorrespondenceStore {
    /// Менеджер подключения к Redis.
    conn: ConnectionManager,
    /// Пространство имён ключей.
    ns: String,
    /// Время жизни записей в секундах.
    ttl_sec: u64,
    /// In-memory LRU-кэш.
    cache: Arc<Mutex<CacheInner>>,
}

impl CorrespondenceStore {
    /// Создаёт хранилище поверх подключения к Redis.
    ///
    /// # Аргументы
    ///
    /// * `conn` — менеджер подключения к Redis.
    /// * `ns` — пространство имён ключей.
    /// * `ttl_sec` — время жизни записей в секундах.
    /// * `cache_max` — максимальный размер in-memory кэша.
    pub fn new(conn: ConnectionManager, ns: String, ttl_sec: u64, cache_max: usize) -> Self {
        CorrespondenceStore {
            conn,
            ns,
            ttl_sec,
            cache: Arc::new(Mutex::new(CacheInner {
                entries: LruCache::new(NonZeroUsize::new(cache_max).unwrap_or(NonZeroUsize::MIN)),
            })),
        }
    }

    /// Возвращает запись соответствия по идентификатору полезной нагрузки.
    ///
    /// Сначала проверяется in-memory кэш; при промахе запись читается из
    /// Redis, парсится из JSON и кладётся в кэш.
    pub async fn get(&self, payload_id: &str) -> Result<Option<CorrRecord>> {
        {
            let mut cache = self.cache.lock().unwrap();
            if let Some((record, written_at)) = cache.entries.get(payload_id) {
                if written_at.elapsed() <= Duration::from_secs(self.ttl_sec) {
                    return Ok(Some(record.clone()));
                }
            }
        }

        let mut conn = self.conn.clone();
        let key = corr_key(&self.ns, payload_id);
        let raw: Option<String> = conn.get(&key).await?;
        let record = match raw {
            Some(json) => {
                let record: CorrRecord = serde_json::from_str(&json).map_err(serde_err)?;
                let mut cache = self.cache.lock().unwrap();
                cache
                    .entries
                    .put(payload_id.to_string(), (record.clone(), Instant::now()));
                Some(record)
            }
            None => None,
        };
        Ok(record)
    }

    /// Сохраняет запись соответствия (write-through: кэш + Redis с TTL).
    pub async fn put(
        &self,
        payload_id: &str,
        original: &str,
        mask: &str,
        types: Vec<String>,
    ) -> Result<()> {
        let record = CorrRecord {
            original: original.to_string(),
            mask: mask.to_string(),
            types,
            created_ts: now_ts(),
        };

        {
            let mut cache = self.cache.lock().unwrap();
            cache
                .entries
                .put(payload_id.to_string(), (record.clone(), Instant::now()));
        }

        let mut conn = self.conn.clone();
        let key = corr_key(&self.ns, payload_id);
        let json = serde_json::to_string(&record).map_err(serde_err)?;
        conn.set_ex::<_, _, ()>(&key, json, self.ttl_sec).await?;
        Ok(())
    }

    /// Идемпотентный поиск соответствия по полезной нагрузке.
    ///
    /// Если запись найдена и `payload` совпадает с оригиналом — возвращает
    /// маску; если `payload` совпадает с маской — возвращает оригинал;
    /// иначе возвращает `None`.
    pub async fn lookup(
        &self,
        payload_id: &str,
        payload: &str,
    ) -> Result<Option<(String, String)>> {
        let record = match self.get(payload_id).await? {
            Some(record) => record,
            None => return Ok(None),
        };
        if payload == record.original {
            Ok(Some((record.mask, record.original)))
        } else if payload == record.mask {
            Ok(Some((record.original, record.mask)))
        } else {
            Ok(None)
        }
    }
}

/// Возвращает текущую временную метку Unix-времени в секундах.
fn now_ts() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

/// Преобразует ошибку сериализации в ошибку Redis.
fn serde_err(e: serde_json::Error) -> redis::RedisError {
    redis::RedisError::from((redis::ErrorKind::ResponseError, "serde", e.to_string()))
}